File: Easy/501. Find Mode in Binary Search Tree/501. Find Mode in Binary Search Tree.py
from collections import Counter

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        def in_order_traversal(node):
            if not node:
                return []

            # Recursively traverse the left subtree
            left_vals = in_order_traversal(node.left)

            # Append the current node's value
            left_vals.append(node.val)

            # Recursively traverse the right subtree
            right_vals = in_order_traversal(node.right)

            return left_vals + right_vals

        # Perform an in-order traversal to get all values in sorted order
        values = in_order_traversal(root)

        # Use Counter to count the occurrences of each value
        value_counts = Counter(values)

        if not value_counts:
            return []

        # Find the maximum count
        max_count = max(value_counts.values())

        # Find the values that have the maximum count
        modes = [
            value for value, count in value_counts.items() if count == max_count
        ]

        return modes

File: Easy/501. Find Mode in Binary Search Tree/501. Find Mode in Binary Search Tree.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn find_mode(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut current_val = 0;
        let mut current_count = 0;
        let mut max_count = 0;
        let mut modes = Vec::new();

        Self::in_order_traversal(
            &root,
            &mut current_val,
            &mut current_count,
            &mut max_count,
            &mut modes,
        );

        modes
    }

    fn in_order_traversal(
        node: &Option<Rc<RefCell<TreeNode>>>,
        current_val: &mut i32,
        current_count: &mut i32,
        max_count: &mut i32,
        modes: &mut Vec<i32>,
    ) {
        if let Some(n) = node {
            let n = n.borrow();

            Self::in_order_traversal(&n.left, current_val, current_count, max_count, modes);

            *current_count = if n.val == *current_val {
                *current_count + 1
            } else {
                1
            };
            *current_val = n.val;

            if *current_count == *max_count {
                modes.push(*current_val);
            } else if *current_count > *max_count {
                *max_count = *current_count;
                *modes = vec![*current_val];
            }

            Self::in_order_traversal(&n.right, current_val, current_count, max_count, modes);
        }
    }
}
